### PR TITLE
Fix alter column for compressed table

### PR DESCRIPTION
**Pull request by RafiaSabih**
Enables adding a boolean column with default value to a compressed table.
This limitation was occurring due to the internal representation of default 
boolean values like 'True' or 'False', hence more checks are added for this.

Fixes https://github.com/timescale/timescaledb/issues/4486

### CODE REVIEW

**Review by svenklemm — APPROVED** [review]

File: src/process_utility.c [hunk 1]
```diff
@@ -282,9 +283,20 @@ check_altertable_add_column_for_compressed(Hypertable *ht, ColumnDef *col)
...
+					 * only allow Const here for now.
+					 */
+					if (IsA(constraint->raw_expr, TypeCast))
+					{
+					   /*
+						* Ignore error for boolean column as values like
+						* 'True' or 'False' are treated as TypeCast.
+						*/
+						char *name = strVal(llast(((TypeCast *) constraint->raw_expr)->typeName->names));
+						is_bool = strcasestr(name, "bool") ? true : false;
```
Comment (akuzm) [inline src/process_utility.c:297]: This feels too specific -- e.g. can I spell it as 'boolean'? What about casts for other types? etc

Maybe instead we could call `eval_const_expressions` before performing the old check?

Comment (svenklemm) [reply src/process_utility.c:297]: This is actually added with that exact spelling by parser. If you look at the tests you can see that the column definition actually uses a different spelling for the column type. This is special casing because the parser treats `'f'` different from `false` or `true` so if you do `ADD COLUMN foo DEFAULT false` it will fail the Const check but `ADD COLUMN foo DEFAULT 'f'` it will be represented as Const. This is specific to bool handling and all the other datatypes should be covered by the Const-check.

Comment (RafiaSabih) [reply src/process_utility.c:297]: Well it doesn't matter what is given in the query, because this value is always bool. TypeName->names is a list of schema and type which is 'pg_catalog' and 'bool' respectively.
Also, so far we only know about this case where we want to allow.

**Comment by mfundul** [comment]
I don't quite understand the commit description. Can you rephrase it somehow?

File: src/process_utility.c [hunk 2]
```diff
@@ -282,9 +283,20 @@ check_altertable_add_column_for_compressed(Hypertable *ht, ColumnDef *col)
 					has_notnull = true;
 					continue;
 				case CONSTR_DEFAULT:
-					/* since default expressions might trigger a table rewrite we
-					 * only allow Const here for now */
-					if (!IsA(constraint->raw_expr, A_Const))
+					/*
+					 * Since default expressions might trigger a table rewrite we
+					 * only allow Const here for now.
```
Comment (mfundul) [inline src/process_utility.c:288]: is this comment about `Const` valid still?

File: src/process_utility.c [hunk 3]
```diff
@@ -282,9 +283,20 @@ check_altertable_add_column_for_compressed(Hypertable *ht, ColumnDef *col)
...
+					if (IsA(constraint->raw_expr, TypeCast))
+					{
+					   /*
+						* Ignore error for boolean column as values like
+						* 'True' or 'False' are treated as TypeCast.
+						*/
+						char *name = strVal(llast(((TypeCast *) constraint->raw_expr)->typeName->names));
+						is_bool = strcasestr(name, "bool") ? true : false;
+					}
+					else if (!is_bool && (!IsA(constraint->raw_expr, A_Const)))
```
Comment (mfundul) [inline src/process_utility.c:299]:
```suggestion
					else if (!is_bool && (!IsA(constraint->raw_expr, TypeCast)))
```
maybe? Not sure.

Comment (mfundul) [inline src/process_utility.c:299]: Isn't `is_bool` always `false` since it's an `else if` so `is_bool` is never set?

File: src/process_utility.c (hunk 1 above)
Comment (svenklemm) [reply src/process_utility.c:297]: To make it bulletproof we should also check that the typecast-arg is a Const, we dont want function calls or anything else here.

File: src/process_utility.c (hunk 3 above)
Comment (svenklemm) [reply src/process_utility.c:299]: yes looks like the else shouldnt be there

File: src/process_utility.c (hunk 1 above)
Comment (svenklemm) [reply src/process_utility.c]: While using `eval_const_expressions` would be the most generic solution it cannot be used here because it only works on planner nodes and the nodes we are dealing with here are parser nodes.

File: src/process_utility.c [hunk 4]
```diff
@@ -282,14 +283,29 @@ check_altertable_add_column_for_compressed(Hypertable *ht, ColumnDef *col)
...
+					 */
 					if (!IsA(constraint->raw_expr, A_Const))
-						ereport(ERROR,
-								(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-								 errmsg("cannot add column with non-constant default expression "
-										"to a hypertable that has compression enabled")));
-
+					{
+						if (IsA(constraint->raw_expr, TypeCast) &&
+							IsA(((TypeCast *) constraint->raw_expr)->arg, Const))
```
Comment (svenklemm) [inline src/process_utility.c:293]:
```suggestion
							IsA((castNode(TypeCast,constraint)->raw_expr)->arg, A_Const))
```

File: src/process_utility.c [hunk 5]
```diff
@@ -282,14 +283,29 @@ check_altertable_add_column_for_compressed(Hypertable *ht, ColumnDef *col)
...
+						{
+						   /*
+							* Ignore error only for boolean column, as values like
+							* 'True' or 'False' are treated as TypeCast.
+							*/
+							char *name =
+									strVal(llast(((TypeCast *) constraint->raw_expr)->typeName->names));
```
Comment (svenklemm) [inline src/process_utility.c:300]:
```suggestion
									strVal(llast((castNode(TypeCast,constraint)->raw_expr)->typeName->names));
```

File: src/process_utility.c (hunk 2 above)
Comment (svenklemm) [reply src/process_utility.c:288]: yes

**Review by mfundul — APPROVED** [review]
